osu.Game/Skinning/RealmBackedResourceStore.cs

```diff
@@ -1,51 +1,78 @@
 using osu.Game.Database;
 using osu.Game.Extensions;
+using Realms;
 
 namespace osu.Game.Skinning
 {
-    public class RealmBackedResourceStore : ResourceStore<byte[]>
+    public class RealmBackedResourceStore<T> : ResourceStore<byte[]>
+        where T : RealmObject, IHasRealmFiles, IHasGuidPrimaryKey
     {
-        private readonly Dictionary<string, string> fileToStoragePathMapping = new Dictionary<string, string>();
+        private Lazy<Dictionary<string, string>> fileToStoragePathMapping;
+
+        private readonly Live<T> liveSource;
+        private readonly IDisposable? realmSubscription;
 
-        public RealmBackedResourceStore(IHasRealmFiles source, IResourceStore<byte[]> underlyingStore, string[] extensions = null)
+        public RealmBackedResourceStore(Live<T> source, IResourceStore<byte[]> underlyingStore, RealmAccess? realm)
             : base(underlyingStore)
         {
-            // Must be initialised before the file cache.
-            if (extensions != null)
-            {
-                foreach (string extension in extensions)
-                    AddExtension(extension);
-            }
+            liveSource = source;
 
-            initialiseFileCache(source);
+            invalidateCache();
+            Debug.Assert(fileToStoragePathMapping != null);
+
+            realmSubscription = realm?.RegisterForNotifications(r => r.All<T>().Where(s => s.ID == source.ID), skinChanged);
         }
 
-        private void initialiseFileCache(IHasRealmFiles source)
+        protected override void Dispose(bool disposing)
         {
-            fileToStoragePathMapping.Clear();
-            foreach (var f in source.Files)
-                fileToStoragePathMapping[f.Filename.ToLowerInvariant()] = f.File.GetStoragePath();
+            base.Dispose(disposing);
+            realmSubscription?.Dispose();
         }
 
+        private void skinChanged(IRealmCollection<T> sender, ChangeSet changes, Exception error) => invalidateCache();
+
         protected override IEnumerable<string> GetFilenames(string name)
         {
             foreach (string filename in base.GetFilenames(name))
             {
-                string path = getPathForFile(filename.ToStandardisedPath());
+                string? path = getPathForFile(filename.ToStandardisedPath());
                 if (path != null)
                     yield return path;
             }
         }
 
-        private string getPathForFile(string filename) =>
-            fileToStoragePathMapping.TryGetValue(filename.ToLower(), out string path) ? path : null;
+        private string? getPathForFile(string filename)
+        {
+            if (fileToStoragePathMapping.Value.TryGetValue(filename.ToLowerInvariant(), out string path))
+                return path;
+
+            return null;
+        }
+
+        private void invalidateCache() => fileToStoragePathMapping = new Lazy<Dictionary<string, string>>(initialiseFileCache, LazyThreadSafetyMode.ExecutionAndPublication);
```

Comment:
As an aside, `ExecutionAndPublication` [is the default](https://github.com/dotnet/runtime/blob/4f66a25e5837f9e6a1edf5b4d88f4ed9c3f0274f/src/libraries/System.Private.CoreLib/src/System/Lazy.cs#L207-L208), so technically it shouldn't be required to specify that.